Fix killer weapon type and icon

menof36go opened the pull request:
Uses local stored weapon entity by class instead of sending entity id. Properly shows the weapon you were killed by and its icon in chat and info popup.

menof36go committed:
Fix killer weapon type and icon

Use local stored weapon instead of sending entire serverside entity

saibotk commented on lua/autorun/client/cl_roundinfo_handler.lua:
```diff
@@ -166,7 +166,7 @@ if CLIENT then
         if (GAMEMODE.round_state ~= ROUND_ACTIVE and GAMEMODE.round_state ~= ROUND_POST) then return end
 
-        local txt = LANG.GetParamTranslation('ttt_rs_suicideText', {killtype = self.data.damage_type_name})
+        local txt = LANG.GetParamTranslation('ttt_rs_suicideText', {killtype = self.data.killer_weapon_name != 'WEAPON_NAME' and self.data.killer_weapon_name or self.data.damage_type_name})
```
I would suggest to rework this, since `self.data.killer_weapon_name != 'WEAPON_NAME'` will misbehave, when a player already got killed by a weapon, as the self.data.killer_weapon_name is never reset to `WEAPON_NAME`

menof36go replied:
I didn't realize that the killer info is reused. Instead of reworking said line of code, I think it would make more sense to make sure that the shared object is set back to its initial values during a call to "HidePopup". That way the table would be cleaned up before it is used again. I think it makes sense to get rid of the leftovers from the last time it was used. What do you think?

menof36go replied:
I implemented the change I proposed and the issue should be resolved. Please let me know what you think.

saibotk reviewed (changes requested):
Please look at my comment above,
also @Alf21 has to take a look.

Thank you for your contribution! :+1:

Alf21 commented on lua/autorun/client/cl_roundinfo_receive.lua:
```diff
@@ -85,8 +85,9 @@ if CLIENT then
 
-		local wep_class = net.ReadEntity()
-		if not IsValid(wep_class) or not wep_class then
+		local hasWeapon = net.ReadUInt(1) == 1
```
Better use net.Read/WriteBit/Bool instead

Alf21 commented on lua/autorun/client/cl_roundinfo_receive.lua:
```diff
@@ -85,8 +85,9 @@ if CLIENT then
-		if not IsValid(wep_class) or not wep_class then
+		local hasWeapon = net.ReadUInt(1) == 1
+
+		if !hasWeapon then
```
Please use the word "not" instead of "!"

Alf21 commented on lua/autorun/client/cl_roundinfo_receive.lua:
```diff
@@ -98,19 +99,16 @@ if CLIENT then
 		end
 
+		local wep_className = net.ReadString()
+		local wep_class = weapons.GetStored(wep_className) or {}
```
Does it really make sense to continue (not return) if the weapon is not found?

menof36go replied:
I restructured it. In the case that no stored weapon with that class name exists we will treat it as if the person was killed with no weapon. Please let me know what you think.

Alf21 replied:
yea that would be the best solution IMO :)

Alf21 commented on lua/autorun/server/sv_roundinfo_send.lua:
```diff
@@ -130,6 +130,7 @@ if SERVER then
 		local wep_class = util.WeaponFromDamage(dmg)
 
 		if not IsValid(wep_class) or not wep_class then
+			net.WriteUInt(0, 1) -- Has Weapon : FALSE
```
Better use net.Read/WriteBit/Bool instead

Alf21 commented on lua/autorun/server/sv_roundinfo_send.lua:
```diff
@@ -157,8 +158,9 @@ if SERVER then
-			
-		net.WriteEntity(wep_class)
+
+		net.WriteUInt(1, 1) -- Has Weapon : TRUE
```
Better use net.Read/WriteBit/Bool instead

menof36go committed:
Reset Killer Info

Reset Killer Info after it was used
Change all ! to not
Change all 1 bit WriteUInt to WriteBool
Change all 1 bit ReadUInt to ReadBool

Alf21 commented on lua/autorun/client/cl_roundinfo_handler.lua:
```diff
@@ -166,7 +172,7 @@ if CLIENT then
         if (GAMEMODE.round_state ~= ROUND_ACTIVE and GAMEMODE.round_state ~= ROUND_POST) then return end
 
-        local txt = LANG.GetParamTranslation('ttt_rs_suicideText', {killtype = self.data.killer_weapon_name != 'WEAPON_NAME' and self.data.killer_weapon_name or self.data.damage_type_name})
+        local txt = LANG.GetParamTranslation('ttt_rs_suicideText', {killtype = not self.data.killer_weapon_name == 'WEAPON_NAME' and self.data.killer_weapon_name or self.data.damage_type_name})
```
`!=` in lua is equal to `~=` (which is prefered :D)

Alf21 commented on lua/autorun/client/cl_roundinfo_receive.lua:
```diff
@@ -85,37 +85,38 @@ if CLIENT then
+					if killer_popup then KILLER_INFO:DisplayPopupKillerWeapon(display_time) end
+					if killer_text then KILLER_INFO:PrintKillerWeapon() end
+				end
+				if killer_type == 2 then
```
`elseif` should work too :D (I think that's an artifact)

Alf21 commented on lua/autorun/client/cl_roundinfo_receive.lua:
```diff
@@ -85,37 +85,38 @@ if CLIENT then
+			if killer_popup then KILLER_INFO:DisplayPopupKillerNoWeapon(display_time) end
+			if killer_text then KILLER_INFO:PrintKillerNoWeapon() end
 		end
 		if killer_type == 2 then
```
`elseif` should work too :D (I think that's an artifact)

menof36go committed:
Remove some artifacts and improve code style

Alf21 commented:
Nice, one little thing: How about using a metatable as a fallback? Then you don‘t need to reset the entire table every time.

menof36go committed:
Add fallback metatable
